File: 287. Find the Duplicate Number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        '''
        https://leetcode.com/articles/find-the-duplicate-number/
        Approach 3: fast slow pointer
        Time O(n)
        Space O(1)
        '''
        fast, slow = nums[0], nums[0]
        while 1:
            fast = nums[nums[fast]]
            slow = nums[slow]
            if fast == slow:
                break
        
        p1 = nums[0]
        p2 = slow
        while p1 != p2:
            p1 = nums[p1]
            p2 = nums[p2]
        return p1

File: 287. Find the Duplicate Number.java
/**
 * Similiar to finding the circle in linkedlist and finding the looping point
 * Time O(n)
 * Space O(1)
 */
class Solution {
    public int findDuplicate(int[] nums) {
        int slow = nums[0], fast = nums[0];
        while (true) {
            slow = nums[slow];
            fast = nums[nums[fast]];
            if (slow == fast) {
                break;
            }
        }
        
        int ptr1 = nums[0], ptr2 = slow;
        while (ptr1 != ptr2) {
            ptr1 = nums[ptr1];
            ptr2 = nums[ptr2];
        }
        return ptr1;
    }
}